File: vscode-extension/src/rules/bugs/unsafeEvalRule.ts
import { Rule, RuleStage } from '../rule';
import { Finding, FindingSeverity, FindingCategory, FindingConfidence } from '../../models/finding';
import { ProjectContext } from '../../scanner/projectContext';
import { findNodesByType, getCallName, isFunctionCall } from '../../ast/traversal';

export class UnsafeEvalRule implements Rule {
  readonly code = 'unsafe-eval';
  readonly stage = RuleStage.ast;

  async evaluate(context: ProjectContext): Promise<Finding[]> {
    const findings: Finding[] = [];

    for (const file of context.files) {
      if (!/\b(?:eval|Function)\s*\(/.test(file.content)) continue;

      const astNode = await context.getAst(file);

      // --- AST path ---
      if (astNode) {
        const calls = findNodesByType(astNode, ['call_expression', 'new_expression', 'expression_statement', 'constructor_invocation']);

        for (const node of calls) {
          if (!isFunctionCall(node)) {
            continue;
          }
          const callName = getCallName(node);
          if (callName === 'eval' || callName === 'Function') {
            findings.push(new Finding({
              category: FindingCategory.bug,
              code: this.code,
              severity: FindingSeverity.medium,
              confidence: FindingConfidence.medium,
              message: `Found direct usage of ${callName}() -> ${node.text.substring(0, 30)}...`,
              fix: `Remove ${callName}() and use safer alternatives like JSON.parse or safe expression parsers.`,
              risk: 'Using dynamic code execution paths is highly dangerous and can allow arbitrary code execution.',
              filePath: file.relativePath,
              line: node.startPosition.row + 1,
              astUsed: true,
            }));
          }
        }
        continue;
      }

      // --- Regex fallback ---
      if (file.astStatus === 'failed') {
        for (let i = 0; i < file.lines.length; i++) {
          const line = file.lines[i];
          const evalMatch = line.match(/\b(eval|Function)\s*\(/);
          if (evalMatch) {
            // Skip if it's in a comment
            const trimmed = line.trimStart();
            if (trimmed.startsWith('//') || trimmed.startsWith('*') || trimmed.startsWith('/*')) continue;

            findings.push(new Finding({
              category: FindingCategory.bug,
              code: this.code,
              severity: FindingSeverity.medium,
              confidence: FindingConfidence.low,
              message: `Found usage of ${evalMatch[1]}() (AST unavailable, regex fallback)`,
              fix: `Remove ${evalMatch[1]}() and use safer alternatives.`,
              risk: 'Dynamic code execution is dangerous. Lower confidence due to regex-only detection.',
              filePath: file.relativePath,
              line: i + 1,
              astUsed: false,
            }));
          }
        }
      }
    }

    return findings;
  }
}

File: lib/src/rules/security/unsafe_eval_rule.dart
import '../../models/finding.dart';
import '../../models/project_context.dart';
import '../../models/scanned_file.dart';
import '../../rule.dart';
import '../rule_helpers.dart';

/// Detects usage of `eval(...)` and `Function.apply(...)` — the only runtime
/// code-evaluation entry points reachable from Dart.
///
/// A bare `Function(...)` expression is NOT flagged: in Dart `Function` is a
/// type, so `void Function()`, `Function() callback`, and `typedef Foo =
/// Function()` are all declarations — not executable code. The older pattern
/// `\bFunction\s*\(` generated thousands of false positives on idiomatic Dart
/// and has been removed.
class UnsafeEvalRule extends Rule {
  const UnsafeEvalRule();

  @override
  String get code => 'unsafe-eval';

  // `eval(` covers `dart:js` / `dart:js_interop` bridges such as
  // `js.context.callMethod('eval', [...])` or a helper named `eval(...)`.
  static final _evalPattern = RegExp(r'''\beval\s*\(''');

  // `Function.apply(...)` is Dart's reflective invocation API. It is NOT a
  // string-eval, but it can still invoke arbitrary closures with arbitrary
  // arguments; flagging it is worthwhile as a medium-confidence signal.
  static final _functionApplyPattern = RegExp(r'''\bFunction\s*\.\s*apply\s*\(''');

  @override
  List<Finding> evaluate(ProjectContext context) {
    final findings = <Finding>[];

    for (final file in context.appDartFiles) {
      _scanPattern(file, _evalPattern, 'eval()', findings);
      _scanPattern(
        file,
        _functionApplyPattern,
        'Function.apply()',
        findings,
      );
    }

    return findings;
  }

  void _scanPattern(
    ScannedFile file,
    RegExp pattern,
    String callName,
    List<Finding> findings,
  ) {
    for (final match in pattern.allMatches(file.content)) {
      final line = file.lineForOffset(match.start);
      if (isCommentLine(file.lines[line - 1])) continue;

      findings.add(
        Finding(
          severity: FindingSeverity.medium,
          confidence: FindingConfidence.medium,
          category: FindingCategory.security,
          code: code,
          message: 'Unsafe $callName call detected',
          fix:
              'Avoid using $callName with dynamic content. Refactor to use '
              'safer alternatives like a lookup table or predefined operations.',
          risk:
              '$callName executes arbitrary code at runtime and can be exploited '
              'for code injection attacks.',
          filePath: file.relativePath,
          line: line,
        ),
      );
    }
  }
}
